src/TesApi.Web/KeyedGroup.cs

```diff
@@ -0,0 +1,162 @@
+﻿// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+using System.Linq;
+
+namespace TesApi.Web
+{
+    /// <summary>
+    /// Provides the abstract base class for a collection whose keys are derivable algorithmically from the values.
+    /// </summary>
+    /// <typeparam name="TElement"></typeparam>
+    /// <typeparam name="TSet"></typeparam>
+    public abstract class KeyedGroup<TElement, TSet> : IEnumerable<TElement> where TSet : GroupableSet<TElement>

```

Comment:
What is the value prop of this over a `KeyedCollection`?

Reply:
It's a not-so-thin wrapper around `KeyedCollection` where it's easier to prevent external manipulations I don't want.

Reply:
Further context: I need to find items using a key of type string, while the contents of the collection are `IBatchPool`, and using a `KeyedCollection` directly with the key type of string breaks using a comparer for IBatchPool (the comparer has to be of the same type as key, and I need the comparer and the key to be of two different types).

Reply:
Thanks for the additional context. It's odd that you have to compare distinct types with only an interface implementation in common. The interface seems like the wrong abstraction here. Interfaces describe behavior, not state - which is required to determine equality. Instead of using an interface, the types in question could share a common base class, and there you can override `equals` to whatever semantics fit your scenario.

Reply:
I'll look at this again once I clear my head from everything else in this PR